pygeos/predicates.py

```diff
@@ -48,6 +49,31 @@ def has_z(geometry, **kwargs):
     return lib.has_z(geometry, **kwargs)
 
 
+@requires_geos("3.7.0")
+@multithreading_enabled
+def is_ccw(geometry, **kwargs):
+    """Returns True if a linestring is counterclockwise.
+
+    Parameters
+    ----------
+    geometry : Geometry or array_like
+        This function will return False for non-linear goemetries and for
+        lines with fewer than 4 points (including the closing point).
```

Comment:
So it doesn't need to be a ring, or not even a "simple" geometry. Using the example from the is_ring docstring (closed but not simple line):

```
In [4]: pygeos.is_ccw(pygeos.Geometry("LINESTRING(0 0, 1 1, 0 1, 1 0, 0 0)"))
Out[4]: True
```

Reply:
It is indeed not enforced that the geometry is "simple", but I think it should. The outcome makes no sense.

This is related to this:
 - https://trac.osgeo.org/geos/ticket/1000

Shall I go ahead and check if a geometry is closed & simple, and if not, return False?

Reply:
This seems problematic too, because (if I understand correctly), this is actually a 3-state problem rather than 2 binary states.

In the binary interpretation, if it is not counterclockwise (`is_ccw` is `False`) that directly implies that it is clockwise (and that inverting coordinates would then make `is_ccw` `True`).  That seems incorrect according to the definition.  If the curve is not simple and closed, it is not [`orientable`](https://en.wikipedia.org/wiki/Curve_orientation).

Instead, it seems like both `is_ccw` and its binary inverse are undefined (or both `False`) when the geometry is not a simple closed curve.  Not sure what to return here in that case, `None`?

Reply:
We have more functions that simple return `False` if they would otherwise error. See for instance `is_ring`: "This function will return False for non-linestrings."

If we would have to allow a third option, then we couldn't use `np.bool`. Resorting to some numeric type `np.int8` would become very confusing (e.g. `bool(np.int8(-1))` evaluates to `True`). `np.object` would simply mean to much overhead.

I don't see much of a problem in returning `False` for unsupported geometries, as long as it is consistent and documented. In that case you would expect users to do the check.

The problem here is, the outcome is not consistent: it is sometimes `True` while it should have errored (-> see my examples in the post below).

Reply:
Indeed, although the property of being able to inverse is_ccw to get is_cw is nice, this will only work for "valid" values for the operation, given our decision to typically return False for invalid values/types instead of raising. 
But so if you ensure that the values are valid (like combining with is_valid/is_simple, as documented now), you can still invert it.